teuthology/orchestra/remote.py

```diff
@@ -449,6 +449,43 @@ def get_tar_stream(self, path, sudo=False):
             ])
         return self.run(args=args, wait=False, stdout=run.PIPE)
 
+    def write_file(self, path, data, sudo=False, mode=None, owner=None,
+                                     mkdir=False, append=False):
+        """
+        Write data to remote file
+
+        :param path:    file path on remote host
+        :param data:    str, binary or fileobj to be written
+        :param sudo:    use sudo to write file, defaults False
+        :param mode:    set file mode bits if provided
+        :param owner:   set file owner if provided
+        :param mkdir:   preliminary create the file directory, defaults False
+        :param append:  append data to the file, defaults False
+        """
+        dd = 'sudo dd' if sudo else 'dd'
+        args = dd + ' of=' + path
+        if append:
+            args += ' conv=notrunc oflag=append'
+        if mkdir:
+            mkdirp = 'sudo mkdir -p' if sudo else 'mkdir -p'
+            dirpath = os.path.dirname(path)
+            if dirpath:
+                args = mkdirp + ' ' + dirpath + '\n' + args
+        if mode:
+            chmod = 'sudo chmod' if sudo else 'chmod'
+            args += '\n' + chmod + ' ' + mode + ' ' + path
+        if owner:
+            chown = 'sudo chown' if sudo else 'chown'
+            args += '\n' + chown + ' ' + owner + ' ' + path
+        args = 'set -ex' + '\n' + args
+        self.run(args=args, stdin=data)
```

Comment:
Oh, I must also mention that I would prefer returning the return value of `self.run()` since access to the object representing the process can allow better testing and troubleshooting. Applies to this method as well as to other methods here and in `misc.py`.

Reply:
As I said, we don't need trouble shoot write_file, we need to stop as fast as possible and analize logs, even more, it is bad to continue execution because it make more troubles to figure out the first source of the failure.

Reply:
So, if we want more control on error handling I would prefer to specific exception produced by the file operation, so the user don't need to check output of the script of write_file. We need to declare all specific to the function exception in the __doc__ for the function and we must not let user to do this instead of us, because the code of script can be changed and will be very difficult to support. Basically user should not use anything about internals of functions like write_file and treat it as blackbox.